File: Leetcode3634. Min removals to balance array/Solution.cpp
class Solution {
public:
    int minRemoval(vector<int>& nums, int k) {
        int n = nums.size();
        sort(nums.begin(), nums.end());

        int mx=0, l=0;
        for(int r=0; r<n; r++) {
            while((l < r) && ((long long)nums[l]*k < nums[r]))
                l++;
            mx = max(mx, r-l+1);
        }
        return n-mx;
    }
};


// runtime - 2ms
const int N=1024, mask=1023, bshift=10;
int freq[N];

void radix_sort(int* nums, int n) {
    int* output = (int*)alloca(n * sizeof(int));  
    unsigned M=*max_element(nums, nums+n);
    const int Mround = max(1, int(31-countl_zero(M)+bshift)/bshift);
    int* in=nums;
    int* out=output;
    bool swapped=0;

    for (int round=0; round<Mround; round++) {
        const int shift=round*bshift;
        memset(freq, 0, sizeof(freq));

        for (int i = 0; i < n; i++)
            freq[(in[i] >> shift) & mask]++;

        partial_sum(freq, freq+N, freq);

        for (int i = n - 1; i >= 0; i--) {
            uint64_t val = in[i];
            uint64_t x = (val >> shift) & mask;
            out[--freq[x]] = val;
        }

        swap(in, out);
        swapped = !swapped;
    }

    if (swapped)
        memcpy(nums, in, n * sizeof(int));
}
class Solution {
public:
    using ll=long long;
    static int minRemoval(vector<int>& nums_, int k) {
        const int n=nums_.size();
        int* nums=nums_.data();
        if (n<=1024) sort(nums, nums+n);
        else radix_sort(nums, n);
        if ((ll)nums[0]*k>=nums[n-1]) return 0;
        int ans=n;
        for(int l=0, r=0; r<n; r++){
            const int x=nums[r];
            while(l<r && (ll)nums[l]*k<x) l++;
            ans=min(ans, n-(r-l+1));
        }
        return ans;
    }
};


// runtime - 3ms
#include <algorithm>
#include <array>
#include <cstdint>
#include <numeric>
#include <span>
#include <utility>
#include <vector>

using u8 = uint8_t;
using u16 = uint16_t;
using u32 = uint32_t;
using u64 = uint64_t;

using i8 = int8_t;
using i16 = int16_t;
using i32 = int32_t;
using i64 = int64_t;

#define FORCE_INLINE inline __attribute__((always_inline))
#define INLINE_LAMBDA __attribute__((always_inline))

template <typename To>
inline static constexpr auto cast = []<typename From>(From&& v) INLINE_LAMBDA
{
    return static_cast<To>(std::forward<From>(v));
};

#define HOT_PATH __attribute__((hot))

#define NO_SANITIZERS \
    __attribute__((no_sanitize("undefined", "address", "coverage", "thread")))

template <typename To, typename From, std::size_t extent = std::dynamic_extent>
    requires(sizeof(To) == sizeof(From))
[[nodiscard]] FORCE_INLINE static auto reinterpret_range(
    std::span<From, extent> in) noexcept
{
    return std::span<To, extent>{
        reinterpret_cast<To*>(in.data()),  // NOLINT
        in.size()};
}

template <typename To, typename From, typename Allocator>
    requires(sizeof(To) == sizeof(From))
[[nodiscard]] FORCE_INLINE static std::span<To> reinterpret_range(
    std::vector<From, Allocator>& v) noexcept
{
    return reinterpret_range<To>(std::span{v});
}

enum class SortOrder : u8
{
    Ascending,
    Descending
};

template <
    std::integral T,
    SortOrder order,
    bool stable,
    u8 bits_per_pass,
    u32 num_passes = ((sizeof(T) * 8 + bits_per_pass - 1) / bits_per_pass),
    u32 capacity = 100'001>
    requires(((num_passes * bits_per_pass) <= sizeof(T) * 8) && (sizeof(T) > 1))
class RadixSorter
{
    using UT = std::make_unsigned_t<T>;

    static constexpr UT base = 1u << bits_per_pass;
    static constexpr UT mask = base - 1;
    static constexpr auto pass_idx_seq =
        std::make_integer_sequence<u8, num_passes>();
    static constexpr size_t num_bits = sizeof(T) * 8;

    inline static std::array<UT, base> count;
    inline static std::array<UT, capacity> temp;

    template <u8 pass_index>
    FORCE_INLINE static void do_pass(std::span<UT> arr) noexcept
        NO_SANITIZERS HOT_PATH
    {
        count.fill(0);
        constexpr UT shift = pass_index * bits_per_pass;
        constexpr UT msb = UT{1} << (num_bits - 1);
        constexpr bool sign_masking = std::is_signed_v<T>;
        constexpr bool is_first_pass = pass_index == 0;
        constexpr bool is_last_pass = (pass_index == (num_passes - 1));
        constexpr UT pre_sign_mask = sign_masking && is_first_pass ? msb : UT{};
        constexpr UT post_sign_mask = sign_masking && is_last_pass ? msb : UT{};

        // Count digit occurrences
        for (auto& v : arr)
        {
            v ^= pre_sign_mask;
            ++count[(v >> shift) & mask];
        }

        const u32 n = cast<u32>(arr.size());

        if constexpr (order == SortOrder::Ascending)
        {
            if constexpr (stable)
            {
                std::inclusive_scan(count.begin(), count.end(), count.begin());

                for (u32 i = n; i--;)
                {
                    UT digit = (arr[i] >> shift) & mask;
                    temp[--count[digit]] = arr[i] ^ post_sign_mask;
                }
            }
            else
            {
                std::exclusive_scan(
                    count.begin(),
                    count.end(),
                    count.begin(),
                    0);

                for (u32 j = 0; j != n; ++j)
                {
                    UT v = arr[j];
                    temp[count[(v >> shift) & mask]++] = v ^ post_sign_mask;
                }
            }
        }
        else  // Descending
        {
            // Compute descending start positions
            std::exclusive_scan(
                count.rbegin(),
                count.rend(),
                count.rbegin(),
                UT{0});

            for (u32 i = 0; i != n; ++i)
            {
                UT v = (arr[i] >> shift) & mask;
                temp[count[v]++] = arr[i] ^ post_sign_mask;
            }
        }

        std::ranges::copy_n(temp.begin(), n, arr.begin());
    }

    // Invokes do_pass for each pass_index.
    template <u8... pass_index>
    FORCE_INLINE static void do_passes(
        std::span<UT> arr,
        std::integer_sequence<u8, pass_index...>) noexcept
        NO_SANITIZERS HOT_PATH
    {
        (do_pass<pass_index>(arr), ...);
    }

public:
    FORCE_INLINE static void sort(std::span<T> arr) noexcept
        NO_SANITIZERS HOT_PATH
    {
        if (arr.size()) do_passes(reinterpret_range<UT>(arr), pass_idx_seq);
    }
};

template <
    u8 bits_per_pass,
    u32 num_passes = 0xFFFFFFFF,
    SortOrder order = SortOrder::Ascending,
    u32 capacity = 100'001,
    std::integral T>
FORCE_INLINE void radix_sort(std::span<T> arr) noexcept NO_SANITIZERS
{
    constexpr u32 np =
        num_passes == 0xFFFFFFFF
            ? ((sizeof(T) * 8 + bits_per_pass - 1) / bits_per_pass)
            : num_passes;
    RadixSorter<T, order, false, bits_per_pass, np, capacity>::sort(arr);
}

template <
    u8 bits_per_pass,
    u32 num_passes = 0xFFFFFFFF,
    SortOrder order = SortOrder::Ascending,
    u32 capacity = 100'001,
    std::integral T>
FORCE_INLINE void stable_radix_sort(std::span<T> arr) noexcept NO_SANITIZERS
{
    constexpr u32 np =
        num_passes == 0xFFFFFFFF
            ? ((sizeof(T) * 8 + bits_per_pass - 1) / bits_per_pass)
            : num_passes;
    RadixSorter<T, order, true, bits_per_pass, np, capacity>::sort(arr);
}

class Solution
{
public:
    [[nodiscard]] static constexpr u32 minRemoval(
        std::vector<u32>& nums,
        u32 k) noexcept
    {
        u32 n = static_cast<u32>(nums.size()), x = 1;

        if (n<128)
            ranges::sort(nums);
        else
            radix_sort<10, 3>(std::span{nums});

        
        u64 lim = 0;
        for (u32 i = ~u32{}, j = 0; j != n; ++j)
        {
            if (nums[j] > lim)
            {
                ++i;
                lim = u64{nums[i]} * k;
            }
            else
            {
                ++x;
            }
        }

        return n - x;
    }

    [[nodiscard]] static u32 minRemoval(std::vector<int>& nums, u32 k) noexcept
    {
        return minRemoval(reinterpret_cast<std::vector<u32>&>(nums), k);
    }
};


// runtime - 4ms
const int N=1024, mask=1023, bshift=10;
int freq[N];

void radix_sort(int* nums, int n) {
    int* output = (int*)alloca(n * sizeof(int));  // buffer
    unsigned M=*max_element(nums, nums+n);
    const int Mround = max(1, int(31-countl_zero(M)+bshift)/bshift);
    int* in=nums;
    int* out=output;
    bool swapped=0;

    for (int round=0; round<Mround; round++) {
        const int shift=round*bshift;
        memset(freq, 0, sizeof(freq));

        for (int i = 0; i < n; i++)
            freq[(in[i] >> shift) & mask]++;

        partial_sum(freq, freq+N, freq);

        for (int i = n - 1; i >= 0; i--) {
            uint64_t val = in[i];
            uint64_t x = (val >> shift) & mask;
            out[--freq[x]] = val;
        }

        swap(in, out);
        swapped = !swapped;
    }

    // If needed, copy back
    if (swapped)
        memcpy(nums, in, n * sizeof(int));
}
class Solution {
public:
    using ll=long long;
    static int minRemoval(vector<int>& nums_, int k) {
        const int n=nums_.size();
        int* nums=nums_.data();
        if (n<=1024) sort(nums, nums+n);
        else radix_sort(nums, n);
        if ((ll)nums[0]*k>=nums[n-1]) return 0;
        int ans=n;
        for(int l=0, r=0; r<n; r++){
            const int x=nums[r];
            while(l<r && (ll)nums[l]*k<x) l++;
            ans=min(ans, n-(r-l+1));
        }
        return ans;
    }
};


// runtime - 5ms
const int N=1024, mask=1023, bshift=10;
int freq[N];

void radix_sort(int* nums, int n) {
    int* output = (int*)alloca(n * sizeof(int));  // buffer
    unsigned M=*max_element(nums, nums+n);
    const int Mround = max(1, int(31-countl_zero(M)+bshift)/bshift);
    int* in=nums;
    int* out=output;
    bool swapped=0;

    for (int round=0; round<Mround; round++) {
        const int shift=round*bshift;
        memset(freq, 0, sizeof(freq));

        for (int i = 0; i < n; i++)
            freq[(in[i] >> shift) & mask]++;

        partial_sum(freq, freq+N, freq);

        for (int i = n - 1; i >= 0; i--) {
            uint64_t val = in[i];
            uint64_t x = (val >> shift) & mask;
            out[--freq[x]] = val;
        }

        swap(in, out);
        swapped = !swapped;
    }

    // If needed, copy back
    if (swapped)
        memcpy(nums, in, n * sizeof(int));
}
class Solution {
public:
    using ll=long long;
    static int minRemoval(vector<int>& nums_, int k) {
        const int n=nums_.size();
        int* nums=nums_.data();
        if (n<=1024) sort(nums, nums+n);
        else radix_sort(nums, n);
        if ((ll)nums[0]*k>=nums[n-1]) return 0;
        int ans=n;
        for(int l=0, r=0; r<n; r++){
            const int x=nums[r];
            while(l<r && (ll)nums[l]*k<x) l++;
            ans=min(ans, n-(r-l+1));
        }
        return ans;
    }
};

File: Leetcode3634. Min removals to balance array/Solution.java
class Solution {
    public int minRemoval(int[] nums, int k) {
        int n = nums.length;
        Arrays.sort(nums);
        int mx = 0, l = 0;

        for (int r = 0; r < n; r++) {
            while (l < r && (long) nums[l] * k < nums[r]) {
                l++;
            }
            mx = Math.max(mx, r - l + 1);
        }

        return n - mx;
    }
}


// runtime - 15ms
class Solution {
    public int minRemoval(int[] nums, int k) {
        nums = sort(nums);
        int i=0, j=0, maxLength = 0;

        while(j < nums.length) {
            if(1.0*nums[j]/nums[i] <= k) {
                maxLength = Math.max(maxLength, j-i+1);
                j++;
            } else {
                i++;
            }
        }

        return nums.length - maxLength;
    }

    int[] sort(int[] nums) {
        int mask = 1023;
        int bits = 10;
        for(int i=0; i<32/bits; i++) {
            nums = countSort(nums, mask, i*bits, mask+1);
        }
        return nums;
    }

    int[] countSort(int[] nums, int mask, int shift, int buckets) {
        int[] count = new int[buckets];
        
        for(int n : nums) {
            count[(n >> shift) & mask]++;
        }

        int start = 0;
        for(int i=0; i<buckets; i++) {
            int temp = count[i];
            count[i] = start;
            start += temp;
        }

        int[] output = new int[nums.length];
        for(int n : nums) {
            int i = (n >> shift) & mask;
            output[count[i]] = n;
            count[i]++;
        }

        return output;
    }
}


// runtime - 16ms
class Solution {
    public int minRemoval(int[] nums, int k) {
        nums = sort(nums);
        int i=0, j=0, maxLength = 0;

        while(j < nums.length) {
            if(1.0*nums[j]/nums[i] <= k) {
                maxLength = Math.max(maxLength, j-i+1);
                j++;
            } else {
                i++;
            }
        }

        return nums.length - maxLength;
    }

    int[] sort(int[] nums) {
        int bits = 10;
        int buckets = 1 << bits;
        for(int i=0; i<32/bits; i++) {
            nums = countSort(nums, buckets, i*bits);
        }
        return nums;
    }

    int[] countSort(int[] nums, int buckets, int shift) {
        int[] count = new int[buckets];
        int mask = buckets-1;

        for(int n : nums) {
            count[(n >> shift) & mask]++;
        }

        int start = 0;
        for(int i=0; i<buckets; i++) {
            int temp = count[i];
            count[i] = start;
            start += temp;
        }

        int[] output = new int[nums.length];
        for(int n : nums) {
            int i = (n >> shift) & mask;
            output[count[i]] = n;
            count[i]++;
        }

        return output;
    }
}


// runtime - 17ms
class Solution {
    public int minRemoval(int[] nums, int k) {
        nums = sort(nums);
        int i=0, j=0, maxLength = 0;

        while(j < nums.length) {
            if(1.0*nums[j]/nums[i] <= k) {
                maxLength = Math.max(maxLength, j-i+1);
                j++;
            } else {
                i++;
            }
        }

        return nums.length - maxLength;
    }

    int[] sort(int[] nums) {
        int bits = 10;
        int buckets = 1 << bits;
        for(int i=0; i<32/bits; i++) {
            nums = countSort(nums, buckets, i*bits);
        }
        return nums;
    }

    int[] countSort(int[] nums, int buckets, int shift) {
        int[] count = new int[buckets];
        int mask = buckets-1;

        for(int n : nums) {
            count[(n >> shift) & mask]++;
        }

        int start = 0;
        for(int i=0; i<buckets; i++) {
            int temp = count[i];
            count[i] = start;
            start += temp;
        }

        int[] output = new int[nums.length];
        for(int n : nums) {
            int i = (n >> shift) & mask;
            output[count[i]] = n;
            count[i]++;
        }

        return output;
    }
}


// runtime - 18ms
class Solution {

    public int minRemoval(int[] nums, int k) {

        int result = 0;

        int min = nums[0];
        int max = nums[0];

        boolean isSorted = true;
        boolean kIsEquals = true;

        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];
            min = getMin(min, num);
            max = getMax(max, num);
            if (num < nums[i - 1]) isSorted = false;
            int dif = abs(num - nums[i - 1]);
            if (dif != k) kIsEquals = false;
        }

        if (max < min * k) return 0;

        if (k == 1 || kIsEquals) return minRemovalK(nums, k, isSorted, min, max);

        while (min * k < max) {

            long different = (long) min * k - max;
            if (0 <= different) break;

            int minRemoveCount = 0;
            int maxRemoveCount = 0;

            int min2 = min * k;
            float max2 = (float) max / k;
            for (int num : nums) {
                if (num == -1) continue;
                if (num < max2) minRemoveCount++;
                if (min2 < num) maxRemoveCount++;
            }

            if (result == 0 && 1000 < nums.length && minRemoveCount == maxRemoveCount) {
                result = minRemoveCount;
                break;
            }

            if (minRemoveCount < maxRemoveCount) {
                int newMin = Integer.MAX_VALUE;
                for (int i = 0; i < nums.length; i++) {
                    int num = nums[i];
                    if (num == -1) continue;
                    if (num == min) {
                        nums[i] = -1;
                        result++;
                    } else newMin = getMin(newMin, num);
                }
                min = newMin;
            } else {
                int newMax = 0;
                for (int i = 0; i < nums.length; i++) {
                    int num = nums[i];
                    if (num == -1) continue;
                    if (num == max) {
                        nums[i] = -1;
                        result++;
                    } else newMax = getMax(newMax, num);
                }
                max = newMax;
            }

        }

        return result;
    }

    private int minRemovalK(int[] nums, int k, boolean isSorted, int min, int max) {

        if (!isSorted) Arrays.sort(nums);

        int pValue = -1;
        int pCount = 0;

        int cValue = -1;
        int cCount = 0;

        int minRemoveCount = 0;
        int maxRemoveCount = 0;

        int min2 = min * k;
        float max2 = (float) max / k;

        for (int num : nums) {
            if (num == pValue) pCount++;
            if (num == cValue) cCount++;
            else {
                cValue = num;
                cCount = 1;
            }
            if (pCount < cCount) {
                pValue = cValue;
                pCount = cCount;
            }
            if (num < max2) minRemoveCount++;
            if (min2 < num) maxRemoveCount++;
        }

        int result = getMin(minRemoveCount, maxRemoveCount);
        return getMin(result, nums.length - pCount);
    }

    private int getMin(int a, int b) {
        return (a < b ? a : b);
    }

    private int getMax(int a, int b) {
        return (a < b ? b : a);
    }

    private int abs(int k) {
        return (k < 0 ? -k : k);
    }

}
